### DISCUSSION TITLE
Possible issues with controller and exiting retroarch

### DISCUSSION THREAD

**Discussion by TheJenious0917** [games-on-whales/gow#24] [Q&A] [answered]
Hey! Great project and I think this is a good start on something I've been looking for for a while. 

I've got a few questions / issues.

First, I'm having trouble getting a controller to pass through to RetroArch. I had some initial trouble with uinput and getting mouse control, but I got that worked out. However, I still am not able to get a controller to pass through. Moonlight client responds to the controller, but once I launch the connection to the stream, controller no longer responds. If there are any logs you might need or if this should be an issue created, let me know. 

Second, I had grabbed the initial version that was just the retroarch container prior to you changing the account and repo and including those other containers. When using your initial image, Moonlight gave me an option to launch both a desktop, and retroarch (though the desktop was only black). Retroarch was working and I could even scan for games and launch them.  However, since changing to running a dedicated xorg container, I only get the desktop option from Moonlight. It connects initially to retroarch, but when exiting the app it stops the retroarch container and dumps to the "desktop" instead of exiting back to the moonlight client itself. 

Again, if some of this is just things to be worked out or if these need to be issues, i can add whatever logs you need.

Thanks!

**Comment by ABeltramo** [comment, marked as answer]
Hello @TheJenious0917, thanks for trying out this project!

> I'm having trouble getting a controller to pass through to RetroArch

Controller passthru doesn't work on our side too, I'll open an issue about it, thanks for reminding us about it!
EDIT: opened, checkout progress at https://github.com/games-on-whales/gow/issues/25

> When using your initial image, Moonlight gave me an option to launch both a desktop, and retroarch (though the desktop was only black)

We are in the middle of changing that, the idea is to always share the Desktop as you would do with a remote PC and then leave to users the option to run any GUI app that they wish to, we bundled RetroArch in the [`docker-compose`](https://github.com/games-on-whales/gow/blob/master/docker-compose.yml) file but we can easily have other apps like Firefox, Steam, etc etc.
The end goal is to have a graphical launcher for apps inside that desktop that will take care of downloading and starting/stopping Docker containers instead of having to mess up with the `docker-compose` file.

> since changing to running a dedicated xorg container, I only get the desktop option from Moonlight

That's intended and it will most likely stay this way in future. You can still override the [`apps.json`](https://github.com/games-on-whales/gow/blob/master/images/sunshine/configs/apps.json) file in order to launch apps from there but then you'll need RetroArch to be installed in the same container that runs Sunshine. 
We would like to avoid that and have one container for one app.

>  It connects initially to retroarch, but when exiting the app it stops the retroarch container and dumps to the "desktop" instead of exiting back to the moonlight client itself.

Yep that's how it works right now, I know it's counter intuitive, but it's just because we are in the middle of building it all.. 
I guess you could add 
```yaml
restart: always
```
to the options of the RetroArch container so that exiting will just restart it instead of stopping the container, but I haven't tried it personally.

I hope this answer your questions, I'm working on adding documentation (you can check it out in the [`docs`](https://github.com/games-on-whales/gow/blob/master/docs/README.md) folder) because I'm aware that right now it's all a little bit confused.

Feel free to join our [Discord server](https://discord.gg/kRGUDHNHt2), we would love to hear more feedback from people that are using it!

**Reply by TheJenious0917** [reply]
ok, that all makes sense. Just wanted to make sure I wasn't missing anything! I know it's still early in the project, so just bringing things up as I find them. 

Cheers!

**Reply by ABeltramo** [reply]
Heads up, latest `master` (5f05651ff4eee6e4f97560b840a69497b429c09b) should have fixed the controller issues! 🎉

**Reply by TheJenious0917** [reply]
Hey, thanks for letting me know. I just was playing around with this and deployed the latest images. Unfortunately, controller still isn't working in my case. This might sound like a dumb question, but does the docker host have to have controller drivers installed or anything like that? Also, my setup is not a straight physical host with docker on it, my host is an ubuntu VM on a proxmox host. Not sure if that changes input drivers or anything since they're virtualized.

**Reply by ABeltramo** [reply]
That's very close to the setup I'm using too, I have a Debian VM on Proxmox. The host shouldn't need any driver since we are managing everything from inside Docker.
Try following the instructions on the [troubleshooting](https://github.com/games-on-whales/gow/blob/master/docs/troubleshooting.md) page about the joypad, maybe you didn't updated the `docker-compose.yml` file?
If you still got issues, can you please open up an issue so that we can move the conversation there?
Thanks for your time!
